bindings/rust/s2n-tls-tokio/src/lib.rs

```diff
@@ -404,19 +363,26 @@ where
     fn poll_shutdown(mut self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<io::Result<()>> {
         ready!(self.as_mut().poll_blinding(ctx))?;
 
-        let status = ready!(self.as_mut().with_io(ctx, |mut context| {
-            context.conn.as_mut().poll_shutdown().map(|r| r.map(|_| ()))
-        }));
+        // s2n_shutdown should not be called again if it errors
+        if self.shutdown_error.is_none() {
+            let result = ready!(self.as_mut().with_io(ctx, |mut context| {
+                context.conn.as_mut().poll_shutdown().map(|r| r.map(|_| ()))
+            }));
+            if let Err(error) = result {
+                self.shutdown_error = Some(error);
+                // s2n_shutdown reading might have triggered blinding again
+                ready!(self.as_mut().poll_blinding(ctx))?;
+            }
+        };
 
-        if let Err(e) = status {
-            // In case of an error shutting down, make sure you wait for
-            // the blinding timeout.
-            self.as_mut().set_blinding_timer(Err(e))?;
-            ready!(self.as_mut().poll_blinding(ctx))?;
-            unreachable!("should have returned the error we just put in!");
-        }
+        let tcp_result = ready!(Pin::new(&mut self.as_mut().stream).poll_shutdown(ctx));
 
-        Pin::new(&mut self.as_mut().stream).poll_shutdown(ctx)
+        let result = if let Some(error) = self.shutdown_error.take() {
+            Err(io::Error::from(error))
+        } else {
+            tcp_result
+        };
+        Ready(result)
```

Comment:
Nit: Line 380 looks a little dense. Do you think something like this would be a little more readable?
```suggestion
        if let Some(error) = self.shutdown_error.take() {
            Ready(Err(io::Error::from(error)))
        } else {
            Ready(tcp_result)
        }
```